pkg/cloud/azure/actuators/machine_scope.go

```diff
@@ -160,26 +161,37 @@ func (m *MachineScope) storeMachineStatus(machine *machinev1.Machine) (*machinev
 	return m.MachineClient.UpdateStatus(machine)
 }
 
-// Persist the machine spec and machine status.
-func (m *MachineScope) Persist() error {
-	if m.MachineClient == nil {
-		return fmt.Errorf("machine client is empty")
+func (s *MachineScope) PersistIfNeeded(currentMachine *machinev1.Machine) error {
+	currentConfig, err := MachineConfigFromProviderSpec(currentMachine.Spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to get machine config: %v", err)
 	}
-
-	latestMachine, err := m.storeMachineSpec(m.Machine)
+	currentProviderStatus, err := v1beta1.MachineStatusFromProviderStatus(currentMachine.Status.ProviderStatus)
 	if err != nil {
-		return fmt.Errorf("[machinescope] failed to update machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
+		return fmt.Errorf("failed to get machine providerStatus: %v", err)
 	}
 
-	_, err = m.storeMachineStatus(latestMachine)
-	if err != nil {
-		return fmt.Errorf("[machinescope] failed to store provider status for machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
+	modifiedMachine := s.Machine.DeepCopy()
```

Comment:
I think the clearest thing would be to pass `ProviderStatus` to this function (i.e., Persist). 

I find the `machine` parameter and `s.Machine` confusing. Which takes precedence? Should the machine object we're trying to modify be `machine` or `s.Machine`? It seems that we're only going to do some work if `ProviderStatus` is different - so we could narrow down to just that. 

Thoughts?

Reply:
currentMachine parameter is the original non mutated object so you can check against it to decided if you want to update and hit the API server (may be naming is not clear)
> It seems that we're only going to do some work if ProviderStatus

This needs to check the machine.Spec, machine.Spec.ProviderSpec, machine.Status and machine.Status.ProviderStatus

Reply:
> This needs to check the machine.Spec, machine.Spec.ProviderSpec, machine.Status and machine.Status.ProviderStatus

OK, so pass those things.